Runtime/Behaviours/CinemachineConfiner2D.cs

```diff
@@ -0,0 +1,413 @@
+#if !UNITY_2019_3_OR_NEWER
+#define CINEMACHINE_PHYSICS_2D
+#endif
+
+using System;
+using System.Collections.Generic;
+using Cinemachine.Utility;
+using UnityEditor;
+using UnityEngine;
+
+namespace Cinemachine
+{
+#if CINEMACHINE_PHYSICS_2D
+    /// <summary>
+    /// An add-on module for Cinemachine Virtual Camera that post-processes
+    /// the final position of the virtual camera. It will confine the virtual
+    /// camera view window to the area specified in the Bounding Shape 2D field
+    /// based on the camera's window size and ratio.
+    /// The confining area is baked and cached at start.
+    /// </summary>
+    [SaveDuringPlay, ExecuteAlways]
+    public class CinemachineConfiner2D : CinemachineExtension
+    {
+        /// <summary>The 2D shape within which the camera is to be contained.</summary>
+        [Tooltip("The 2D shape within which the camera is to be contained.  " +
+                 "Can be a 2D polygon or 2D composite collider.")]
+        public Collider2D m_BoundingShape2D;
+
+        /// <summary>Damping applied automatically around corners to avoid jumps.</summary>
+        [Tooltip("Damping applied around corners to avoid jumps.  Higher numbers are more gradual.")]
+        [Range(0, 5)]
+        public float m_Damping = 0;
+        private bool m_CornerDampingIsOn = false;
+        private float m_CornerDampingSpeedup = 1f;
+        private float m_CornerAngleTreshold = 10f;
+
+        /// <summary>Draws Gizmos for easier fine-tuning.</summary>
+        [Tooltip("Draws Input Bounding Shape (black) and Confiner (cyan) for easier fine-tuning.")]
+        public bool m_DrawGizmos = true;
+        [HideInInspector, SerializeField] internal bool m_AutoBake = true; // TODO: remove
```

Comment:
I'm uncomfortable exposing the whole baking thing.  I think that publicly there should be a single concept of a cache.  It's either valid or it's not.  We don't know what gets cached, but it doesn't matter.

If the cache is invalid, it will be automatically recomputed at the next usage (lazy evaluation).

The cache is automatically invalidated in some well-defined circumstances: 
- any relevant parameter in the Confiner changes
- transform of the bounding shape changes (maybe not!  see my other comment)

The cache is NOT automatically invalidated (due to high computation cost every frame) if the contents of the confining shape change (e.g. points get moved dynamically).  In that case, we expose an API to forcibly invalidate the cache so that it gets auto-recomputed next time it's needed.

I think that's easy enough for our users to grasp, and is sufficient for our needs.

The confiner inspector can regularly invalidate the cache, so that changes made to the confiner shape while building it get processed automatically.

Reply:
I agree, and will update the code to reflect this. 

But, I would not invalidate the cache. It can take some time to recompute the confiner cache (several ms).
The user can force rebake through the exposed API.

Reply:
If it's that slow to recompute then maybe we shouldn't rely on lazy evaluation.  Maybe the baked cache should get saved with the scene.